### PR TITLE
Guid windows 20190307

### PR DESCRIPTION
**Pull request by yomimono**
This is a rebase of #32 which I *think* incorporates all of the changes, and is automatically mergeable.  Sorry for the `and yomimono` on all of the commits now, I'm not sure how to get rid of that :(

### CODE REVIEW

File: lib/gUI.ml [hunk 1]
```diff
@@ -19,80 +19,144 @@ let gui_agent_port =
...
-      f "client connected (screen size: %ldx%ld depth: %ld mem: %ldx)"
-        screen_w screen_h xdepth xmem);
-  Lwt.return gui
+type event =
+  | UNIT of unit (* placeholder for unimplemented events *)
+  | Keypress of msg_keypress_t
+  | Focus of msg_focus_t
+  | Motion of msg_motion_t
+  | Clipboard_request
+  | Clipboard_data of Cstruct.t
```
Comment (cfcs) [inline lib/gUI.ml]: `  | Configure of Formats.GUI.msg_configure_t`

File: lib/gUI.ml [hunk 2]
```diff
@@ -19,80 +19,144 @@ let gui_agent_port =
...
-let _decode_MSG_MOTION buf =
-  match Formats.GUI.decode_msg_motion buf with
-  | Some m ->
-    Log.warn (fun f -> f "Motion event: x: %d y: %d state: %ld is_hint: %d"
-                 m.x m.y m.state m.is_hint)
-  | None ->
-    Log.warn (fun f -> f "attempted to decode a motion event, but we were not successful")
+  Log.warn (fun f -> f "Event: CLIPBOARD_DATA: %a" Cstruct.hexdump_pp buf);
+  Clipboard_data buf
+
```
Comment (cfcs) [inline lib/gUI.ml]: hmm somewhere around here it's missing
```ocaml
let decode_MSG_MOTION buf =
  match Formats.GUI.decode_msg_motion buf with
  | Some m ->
    Log.warn (fun f -> f "Motion event: x: %d y: %d state: %ld is_hint: %d"
                 m.x m.y m.state m.is_hint);
    Motion m
  | None ->
    Log.warn (fun f -> f "attempted to decode a motion event, but we were not successful: %a" Cstruct.hexdump_pp buf);
    UNIT ()


let decode_CONFIGURE buf =
  match decode_msg_configure buf with
  | Some m -> Configure m
  | None ->
    Log.warn (fun f -> f "failed decoding CONFIGURE message from dom0: %a"
                 Cstruct.hexdump_pp buf) ;
    UNIT ()
```

File: lib/gUI.ml [hunk 3]
```diff
@@ -19,80 +19,144 @@ let gui_agent_port =
...
-let rec listen t =
-  QV.recv t >>= function
+let rec listen t () =
+  QV.recv t.qv >>= function
   | `Eof -> failwith "End-of-file from GUId in dom0"
   | `Ok (msg_header , msg_buf) ->
-  let msg_window = get_msg_header_window msg_header |> Int32.to_int in
+  let window = get_msg_header_window msg_header in
+  let send_to_window promise =
+    promise >>= fun resolved ->
```
Comment (cfcs) [inline lib/gUI.ml]: in the old branch this one was like:
```ocaml
  let window = get_msg_header_window msg_header in
  let send_to_window event =
    match List.find (fun t -> t.no = window) t.mvar with
    | w -> Lwt_mvar.put w.mvar event
    | exception _ -> Log.warn (fun m -> m "No such window %ld" window);
                     Lwt.return ()
  in
  let msg_len    = get_msg_header_untrusted_len msg_header |> Int32.to_int in
  send_to_window
  begin match int_to_msg_type (get_msg_header_ty msg_header) with
```

File: lib/gUI.ml [hunk 4]
```diff
@@ -102,57 +166,71 @@ let rec listen t =
          | MSG_EXECUTE | MSG_WMNAME | MSG_KEYMAP_NOTIFY | MSG_WINDOW_HINTS
          | MSG_WINDOW_FLAGS | MSG_WMCLASS | MSG_CLIPBOARD_REQ
          | MSG_CLOSE as msg)
-      when msg_len <> (match msg_type_size msg with Some x -> x | None -> -1) ->
+    when (begin match msg_type_size msg with Some x -> x <> msg_len
+                                           | None -> false end) ->
```
Comment (cfcs) [inline lib/gUI.ml]: this `None -> false` should be `None -> true`, I think that was fixed in a commit that got lost somewhere.

I also changed it to not have `Lwt.return` in all the righthand cases here.

File: lib/gUI.mli [hunk 5]
```diff
@@ -4,7 +4,54 @@
...
+
+open Formats.GUI
+
+type event =
+  | UNIT of unit
+  | Keypress of Formats.GUI.msg_keypress_t
+  | Focus of msg_focus_t
+  | Motion of msg_motion_t
+  | Clipboard_request
+  | Clipboard_data of Cstruct.t
```
Comment (cfcs) [inline lib/gUI.mli]:
```ocaml
  | Configure of Formats.GUI.msg_configure_t
```

File: lib/gUI.ml [hunk 6]
```diff
@@ -19,80 +19,144 @@ let gui_agent_port =
...
+  | Window_close
+  | Button of msg_button_t
+
+let pp_event fmt event =
+  let pf() = Format.fprintf fmt in
+  match event with
+  | UNIT () -> pf() "UNIT"
+  | Button _ -> pf() "Button"
+  | Clipboard_request -> pf() "Clipboard_request"
+  | Clipboard_data cs -> pf() "Clipboard_data: %S" (Cstruct.to_string cs)
```
Comment (cfcs) [inline lib/gUI.ml]:
```ocaml
  | Configure x -> pf() "Configure: @[x=%ld;@ y=%ld;@ width=%ld;@ height=%ld@]"
                     x.x x.y x.width x.height
```

**Comment by yomimono** [comment]
Superseded by a less broken PR.

**Comment by talex5** [comment]
Can we delete this branch now?

**Comment by yomimono** [comment]
Yes.

**Pull request closed without merging**